### ISSUE TITLE
Create/define a plugin SDK

### ISSUE THREAD

**Issue by rsmmr** [zeek/zeek#266]
We don't have a well-defined API yet that plugins can use to interface with the rest of Zeek, making it hard to understand what's safe to use from inside a plugin, and also to maintain compatibility across releases.

**Comment by timwoj** [comment]
I'm going to try to kick off the discussion of this, hopefully to get something in place for maybe the 4.1 release.

1. What does a release of an SDK look like? Does the standard Zeek installation only include the bare headers needed to build a plugin and all of the other symbols are simply available via the linker? Zeek most definitely isn't header-only, so it might not be necessary to continue installing all of the headers all of the time. Maybe installing the full set of headers could be a configure-time option?
2. What exactly does the SDK look like? A lot projects follow a model where there is an outward-facing namespace (`zeek`) and an internal one (`zeek::internal`). Anything in the `zeek` namespace is known to be stable outside of our current deprecation system, while anything in the `zeek::internal` namespace can change whenever and shouldn't be relied on to be stable.
3. What should we include in each of the namespaces? Are we thinking just the things that would be required to build plugins such as the input/logging/iosource frameworks? Do we need to include `util.h`, etc? I have a feeling this question is probably the large bulk of the work here and the answer may require restructuring a bunch of code to split parts that should be internal from parts that aren't? There's a lot of global variables that might be needed in the SDK.

**Comment by jsiwek** [comment]
> I'm going to try to kick off the discussion of this

Thanks!

> 1. What does a release of an SDK look like? Does the standard Zeek installation only include the bare headers needed to build a plugin and all of the other symbols are simply available via the linker? Zeek most definitely isn't header-only, so it might not be necessary to continue installing all of the headers all of the time.

My guess is also that it's easiest to install all headers.  Don't think there's ever much contention on minimizing install or anything like that, but the concern has usually been about what bits of those headers users ought to feel safe using in their own plugins, and that's what the following point is about:

> 2. What exactly does the SDK look like? A lot projects follow a model where there is an outward-facing namespace (`zeek`) and an internal one (`zeek::internal`). Anything in the `zeek` namespace is known to be stable outside of our current deprecation system, while anything in the `zeek::internal` namespace can change whenever and shouldn't be relied on to be stable.

Yeah, that's what I think, too.  Everything inside a namespace `zeek` (public API + name collision avoidance) and then something like `zeek::internal` or `zeek::detail` for things we think ought to be implementation details and subject to change at any time.  i.e. "you can use it if you want, but don't expect stable interfaces".

> 3. What should we include in each of the namespaces? Are we thinking just the things that would be required to build plugins such as the input/logging/iosource frameworks? Do we need to include `util.h`, etc? I have a feeling this question is probably the large bulk of the work here and the answer may require restructuring a bunch of code to split parts that should be internal from parts that aren't? There's a lot of global variables that might be needed in the SDK.

Right that this is likely the bulk of open-questions: "what is public vs. private" and "how to re-organize things that get split up".  General thoughts:

* Starting with `util.h` maybe not the best -- have a feeling most of it is fine to be public, but maybe also a few bits that really aren't too useful and won't know until going in for closer look

Most of these are likely desirable as public:

* All the singletons/managers (or generally maybe other global state, too)
* All things meant to be used directly as plugins: input, logging, iosource, analyzers
* Likely all the "frameworks", like stuff that's not directly a plugin itself, but may be used from inside one: supervisor, broker, maybe signatures, maybe zeekygen
* Things needed to interact with the script-layer like Event, Func, ID, Type, Val and maybe other things they end up depending upon (but think most bits related to parsing or traversing AST could be private like Var, Stmt, Expr, Traverse)

**Comment by timwoj** [comment]
> "how to re-organize things that get split up"

I'd say maybe just move the internal stuff to an `internal` directory and leave the rest in the regular source dir. I don't think much more needs to be done than that.

> most bits related to parsing or traversing AST could be private like Var, Stmt, Expr, Traverse

This would be an easy starting point for sure. Once that stuff has been moved then we could review the rest of it.

**Comment by timwoj** [comment]
A starting list for things to be `zeek::internal`:
- [ ] `Stmt`
- [ ] `Frame`
- [ ] `Expr`
- [ ] `Var`
- [ ] `Traverse`
- [ ] `Trigger` (which uses Stmt/Frame/Expr, and probably shouldn't be used by anything external)
- [ ] All of the debugger code

Anything else I should add to the list? I'm definitely not looking to start this work yet, but

**Comment by rsmmr** [comment]
Overall I think we should try to keep the public API as small as we can; the less is in there, the more flexibility we have in changing stuff. To that end, it might be helpful to look at the from the perspective of the various kinds of plugins: If I want to write a new protocol analyzer / log writer / etc, what pieces do I *typically* need to access to achieve that task.

**Comment by jsiwek** [comment]
Was having intrusive thoughts about this, so putting out an attempt to summarize/break-down and propose concrete examples for moving forward.

First thing is to clearly (re)state/define the problem we want to solve: we currently don't have a formal process/guarantee on API stability.  This causes two challenges:

* Users don't know what they're signing up for in terms of stability/maintenance-burden when looking around for bits of Zeek they may want to use in their plugins.
* Zeek maintainers have to go through a cumbersome decision-tree for *any* breaking API change to determine whether it's safe to go ahead with the breakage or to do a formal deprecation process.

If there's other problems we want to solve, please help bring those to attention, but that summary is at least my only complaint about the current state of things.

The solution to formalizing how we signal API stability should be straightforward: put things in separate namespaces.  That clearly communicates to users what to expect and makes the maintainer decision process for breaking API changes much easier since we've offloaded the difficult thought process to an earlier point in time (now) and only have to answer "which namespace is it in?".

We can wordsmith about what names to use for namespaces, but going with `zeek` and `zeek::internal` for following points:

* We've been using "public" vs. "private" terminology, but that's probably misleading since we tend to think of that in relation to what people can and cannot use.  The `internal` namespace wouldn't be restricting what people use -- the namespaces are only going to exist to formalize API stability expectations and `internal` actually means something like "unstable" or "volatile".

* There's not a single "right" way to organize things and it's ok to make "mistakes" during initial re-organization.  We're going to move things to `internal` based on what the Plugin API expects people to use, research what existing plugins currently use, or just make an educated guess.  If we are too conservative and put something in `internal` that users want better API stability for, we can always discuss that with them and move things between namespaces later.

* It would be nice to do this re-namespacing incrementally, rather than all at once.  I expect that makes review/merging/development processes easier and it doesn't cause any more problem than currently exists.  If a release happens and some things still live in global namespace, that just means the API stability guarantees are unknown for those things -- the same situation we're in now.

* It would be nice to do the re-namespacing in a non-breaking way via use of deprecation attributes.  That also better allows for incremental approach and alleviates pressure/timeline issues that would be associated with trying to do one, large breaking change that moves everything into namespaces. Examples:

Putting a global variable in a namespace:

```diff
diff --git a/src/Val.h b/src/Val.h
index 1bcc5cc14..5e3293ad8 100644
--- a/src/Val.h
+++ b/src/Val.h
@@ -470,7 +470,11 @@ private:
 	IntrusivePtr<Val> b_false;
 };
 
+[[deprecated("Remove in 4.1.  Use zeek::val_mgr instead.")]]
+extern ValManager*& val_mgr;
+namespace zeek {
 extern ValManager* val_mgr;
+}
 
 #define Microseconds 1e-6
 #define Milliseconds 1e-3
diff --git a/src/main.cc b/src/main.cc
index 460d6e9b2..88d272c70 100644
--- a/src/main.cc
+++ b/src/main.cc
@@ -84,7 +84,8 @@ int perftools_profile = 0;
 
 DNS_Mgr* dns_mgr;
 TimerMgr* timer_mgr;
-ValManager* val_mgr = nullptr;
+ValManager* zeek::val_mgr = nullptr;
+ValManager*& val_mgr = zeek::val_mgr;
 logging::Manager* log_mgr = nullptr;
 threading::Manager* thread_mgr = nullptr;
 input::Manager* input_mgr = nullptr;
```

Putting a class/type in a namespace:

```diff
diff --git a/src/Val.h b/src/Val.h
index 1bcc5cc14..5b254a821 100644
--- a/src/Val.h
+++ b/src/Val.h
@@ -25,6 +25,10 @@
 #define UDP_PORT_MASK	0x20000
 #define ICMP_PORT_MASK	0x30000
 
+namespace zeek {
+class AddrVal;
+}
+
 template<typename T> class PDict;
 class IterCookie;
 
@@ -36,7 +40,8 @@ class BroFile;
 class PrefixTable;
 
 class PortVal;
-class AddrVal;
+using AddrVal [[deprecated("Remove in 4.1.  Use zeek::AddrVal isntead.")]]
+              = zeek::AddrVal;
 class SubNetVal;
 
 class IntervalVal;
@@ -526,6 +531,7 @@ protected:
 	IntrusivePtr<Val> DoClone(CloneState* state) override;
 };
 
+namespace zeek {
 class AddrVal final : public Val {
 public:
 	explicit AddrVal(const char* text);
@@ -544,6 +550,7 @@ public:
 protected:
 	IntrusivePtr<Val> DoClone(CloneState* state) override;
 };
+}
 
 class SubNetVal final : public Val {
 public:
```

Things we move into `zeek::internal` could work the same way if we want, but can always use Good Judgement if there's things we do not expect anyone to be using currently and want to avoid doing the extra deprecation work for those.

If we choose to move header/source files around to separate out the "internal" stuff in the directory structure itself, this would be an example of how to deprecate the old header file, taken from how Broker's `bro.hh` became `zeek.hh`:

```c++
#pragma once
#pragma message("Warning: bro.hh header is deprecated, use zeek.hh instead")
#include "broker/zeek.hh"
```

**Comment by rsmmr** [comment]
Nice summary of goals and way forward, sounds all good to me.

On the namespace naming, we could use `zeek::detail` instead of `zeek::internal` because (a) it feels like it suggests a bit more wiggle room for users to use if they insist ("internal" sounds like "hands off!" to me, vs. "detail" more like " it's implementation details and we won't guarantee anything if you insist to use it"; and (2) Broker/CAF are already using "detail". That said, I'd be fine either way, no strong feelings.

@timwoj I might be useful to write up our policy on the new namespaces upfront before any work starts, so that we can more easily see if in practice things align with out intent.

**Comment by timwoj** [comment]
> intrusive thoughts

😆 

I agree, great writeup that covers pretty much everything. I agree with making it `zeek::detail`.

> write up our policy on the new namespaces

Adding this section to our Coding Style document on the wiki:

> Zeek Namespaces
> ===============
> 
> Zeek uses two primary namespaces: ``zeek`` and ``zeek::detail``. The ``zeek``
> namespace is intended for code that is used by external plugins and other code
> that expect a stable API. Any changes to this code will strictly follow the
> Zeek team's deprecation process. On the other hand, code in the
> ``zeek::detail`` namespace will not necessarily maintain a stable API. It is
> still available for use by external code, but the APIs included should be not
> be relied upon to never change from release to release.

**Comment by rsmmr** [comment]
+1 as well. I think we should also say that we are in transition to that scheme, and how we are handling that process. How about this externded version, taking some ideas from Jon's write-up:


> Zeek is the process of migration two using two primary namespaces: `zeek` and `zeek::detail`.  The `zeek` namespace is intended for interfaces that are used by external plugins and other code that expect a stable API. Any changes to this code will strictly follow the Zeek team's deprecation process. On the other hand, code in the `zeek::detail` namespace will not necessarily maintain a stable API. It is still available for use by external code, but the APIs included should be not be relied upon to not change from release to release.
> 
> All of Zeek's headers will eventually live inside these two namespaces. As we are moving towards that world, we will follow our standard deprecation process for everything moving into `zeek`, meaning that we will initially keep aliases available inside the current global namespace. We will generally not keep aliases for anything moving into `zeek::internal`, although we may make exceptions in cases of substantial impact on external code. During the transition, everything still living only in the global namespace continues to follow the current semantics: No stability guarantees, but a "best effort" approach of not needlessly breaking APIs between releases.

**Comment by timwoj** [comment]
This is a somewhat naive approach and almost certainly misses things, but it's a starting point. This is a list of the files included by any package with a `build_command` option in the package repo: https://gist.github.com/timwoj/9784367268fbbcda758d73e723eacc2f.

There's obviously an include tree that goes off of this, and it's only a check of the actual included files and not the classes used. Like I said it's a starting point for looking at what is actually used in external code.

**Comment by rsmmr** [comment]
That *is* nice starting point. Interesting to see `analyzer/protocol` & 'file_analysis/analyzer/*` in there, I hadn't thought about whether analyzers should be part of the public API so that people can rely on putting them to use in their own code. I'm torn. Maybe the analyzers themselves could go into public (their API is mostly depending on their parent classes anyways), but any further helper types/functions they define, remain in `detail`?

**Comment by rsmmr** [comment]
One more thought: As we migrate things into namespaces, it would be nice if a plugin author could easy test if they are using only "approved" stuff during the period where still have the aliases for the original names in place. Would it make sense to declare the aliases in the global namespace with a macro that could be turned off through some global preprocessor option? Then plugin authors could set that now already if they want to make sure to really just use the new names. And at the same time, that would give us a way to easily test plugins for things we are breaking by moving something into `detail`.

**Comment by timwoj** [comment]
I added a macro on my branch yesterday to both forward-declare and add an alias:

```
#define FORWARD_DECLARE_NAMESPACED(cls, ns) \
        namespace ns { class cls; } \
        using cls [[deprecated("Remove in v4.1. Use " #ns "::" #cls " instead.")]] = ns::cls;
```

Adding an extra bit in there to turn the aliasing off with a compiler flag is easy enough. Once we're to the point that we can remove the deprecations, this macro can be removed as well and the namespace blocks combined into single blocks.

**Comment by jsiwek** [comment]
> That _is_ a nice starting point. Interesting to see `analyzer/protocol` & `file_analysis/analyzer/*` in there

Probably worth looking closer to see how they use them.  They may be used in a "shallow" way (instantiate, add to tree, or feed data via common Analyzer interface) or just include headers as some copy-paste relict, but don't rely on anything from them.

> I hadn't thought about whether analyzers should be part of the public API so that people can rely on putting them to use in their own code.

My suggestion is to put most of them in `zeek::detail` except if there's ones that stand out as solely there for re-use in other analyzers (e.g. might expect ZIP to go into `zeek`).

Anything in `detail` doesn't mean people can't use or "rely" on them, just they might expect to write and maintain more version-dependent code if they use more than the common Analyzer interface.  Also, if there's demand to "stabilize" some analyzer's API that we initially put in `detail`, I'd think we'd be open to discuss and move it to `zeek` later, too.

> One more thought: As we migrate things into namespaces, it would be nice if a plugin author could easy test if they are using only "approved" stuff during the period where still have the aliases for the original names in place. Would it make sense to declare the aliases in the global namespace with a macro that could be turned off through some global preprocessor option? Then plugin authors could set that now already if they want to make sure to really just use the new names.

Don't really have anything against such a macro, but also don't understand what it helps.

* For things we move into `zeek`, the global alias will come with a deprecation warning saying to use `zeek::`.

* For things we move into `zeek::detail`, there's either no global alias and their plugin no longer builds or there's a global alias with a deprecation warning saying to use `zeek::detail::`.

Either way, they shouldn't need a way to "turn off global aliases" to determine whether that thing they were using is going to fall in the "stable" API or the "volatile" API since they find that out already from reading compiler warning/error messages?

**Comment by timwoj** [comment]
> Probably worth looking closer to see how they use them. 

- The x509 file analysis plugin is used by `0xxon/zeek-plugin-roca/src/roca.bif` and `0xxon-cve-2020-0601-plugin/src/mscve.bif` to inspect certificate data.
- The TCP and UDP protocol analyzers make sense for plugins that add protocols that extend from them, such as  amzn/zeek-plugin-bacnet which adds a new `TCP_ApplicationAnalyzer`.
- The HTTP analyzer is used by multiple plugins to add protocol parsing for HTTP-based protocols, such as mitre/icap.
- mitrecnd/bro-http2 uses ZIP for the implementation, but also includes PIA which is unused.

There's more, but that's a few examples. I have a feeling there's probably places where plugins included a header, and then stopped using it for whatever reason but never removed it. There should be a common entry-point for every plugin (TCP.h, UDP.h, etc) and maybe a few utility-type files (ContentLine.h) that should be in `zeek`. Beyond that, should everything else in all of the protocols just be in `zeek::detail`?

**Comment by timwoj** [comment]
`Frame` and `Func` are used by `Plugin.h` for the `HookCallFunction`, `HookArgument`, `AsFrame`, and `AsFunc` methods.

**Comment by timwoj** [comment]
As an aside, should we use this effort to take some time to clean up the global namespace (or well, global things now stuck into namespaces) while we're at it? For example, I'm looking at `Var.h` right now which is just a bunch of `extern`'d functions thrown in the global namespace. Should I consider moving those into a class as a collection of static methods or a separate namespace inside of `zeek::detail` while I'm reworking all of this? The same goes for places like `Net.h`, etc.

**Comment by rsmmr** [comment]
@jsiwek On the macro: yeah, sounds right, the deprecation warnings should do the trick already.

**Comment by rsmmr** [comment]
> As an aside, should we use this effort to take some time to clean up the global namespace

I'd say yes;  a couple of sub-namespaces would be helpful to group some things, currently it there's really nothing that tells what functions relate to each other.

> After further digging, I'm not sure that's really feasible at least for Var.h

Maybe this needs to be split now into things part of the public API (though it's ironic that those functions are called `internal_*` :-); and private stuff that could move into a sub-namespace of detail.

(I wouldn't just hide functions inside a class if the sole purpose of that is grouping them together; seems namespaces solve that problem already.)

**Comment by jsiwek** [comment]
> > After further digging, I'm not sure that's really feasible at least for Var.h

Coincidentally, I was just on the path to "cleaning up" some of `Var.h` and `NetVar.h` in relation to `IntrusivePtr` migration.  I'm not sure how to best coordinate here -- maybe assume anything that uses a pointer to `BroObj` derivative like `Val*` or `BroType*` will be something I'll have already deprecated and replaced with better named/organized version?

> Maybe this needs to be split now into things part of the public API (though it's ironic that those functions are called `internal_*` :-); and private stuff that could move into a sub-namespace of detail.

Yeah, for the stuff in `Var.h` in particular, I was heading in the direction of separating out the commonly used "lookup" functions from the stuff that's actually details of the parsing process.  Here's notes I made on naming ideas:

```c++
// Functions to "lookup" variables.
const IntrusivePtr<ID>& zeek::var::id(std::string_view name)
const IntrusivePtr<BroType>& zeek::var::type(std::string_view name)
const IntrusivePtr<Val>& zeek::var::val(std::string_view name)
const IntrusivePtr<Val>& zeek::var::const(std::string_view name)
IntrusivePtr<Func> zeek::var::func(std::string_view name)

// Commonly used variable types.
IntrusivePtr<RecordType> zeek::var::conn_id;
...

// Initialize *commonly* used types, like conn_id, other niche types should
// be retrieved via lookup (and maybe stored in local static) closer to usage site.
void zeek::var::detail::init();
```

A separate question I have is if there's any opinions on header-naming-convention?  I was going to put the above in a `ZeekVars.h`.

> (I wouldn't just hide functions inside a class if the sole purpose of that is grouping them together; seems namespaces solve that problem already.)

That was the thought I also arrived at -- wrapping some of the globals in class/struct didn't seem more helpful than just moving them into sub-namespace.  e.g. wrapping in class adds awkwardness of how to do the `detail` part of the interface like the `zeek::var::detail::init()` above -- probably wouldn't want to put that in class as protected/private and add friends, so will just end up being a free function in a `detail` namespace anyway.

**Comment by timwoj** [comment]
> Coincidentally, I was just on the path to "cleaning up" some of Var.h and NetVar.h in relation to IntrusivePtr migration.

I can just shelve my Var stuff until yours lands and take another look after that.

> Yeah, for the stuff in Var.h in particular, I was heading in the direction of separating out the commonly used "lookup" functions from the stuff that's actually details of the parsing process. 

I'd agree with that.

> A separate question I have is if there's any opinions on header-naming-convention? I was going to put the above in a ZeekVar.h.

I have no real hard preference. I think adding `Zeek` to a bunch of file names might be a bit noisy though.

**Comment by jsiwek** [comment]
> I think adding `Zeek` to a bunch of file names might be a bit noisy though.

Was thinking a "nice" thing is in context of external/plugin code readable (or even name-collision-avoidance).  If they have `#include "Var.h"`, is that Zeek's `Var.h`, some random library they're using that also has a `Var.h`, or their own `Var.h`.

We could alternatively move things into separate directory trees entirely:

```
<repo>/include/zeek/Var.h
<repo>/include/zeek/supervisor/Supervisor.h
...
<repo>/src/zeek/Var.cc
<repo>/src/zeek/supervisor/Supervisor.cc
```

Or

```
<repo>/src/zeek/Var.h
<repo>/src/zeek/supervisor/Supervisor.h
...
<repo>/src/Var.cc
<repo>/src/supervisor/Supervisor.cc
```

Such that you end up with `#include "zeek/Var.h`

* Not sure if we'd then also want subdirectories for the `detail` bits
* Also not sure about capitalization -- `Var.h` feels like it should be `var.h` since there's no single, top-level `Var` class/struct as the subject of the files, just various things inside `zeek::var(::detail)` namespace.

**Comment by jsiwek** [comment]
Minor update just for the `Var.h` / `NetVar.h` re-org: I actually went with moving the common type pointers and lookup functions into `ID.h` within a `zeek::id` namespace.  Think that generally makes more sense in relation to actual hierarchy of the terminology: they're operations/references to "identifiers" (which also help refer to "types"), not necessarily just the more-specific "variables" (which mostly help refer to "values").

Bet the general question of header/directory naming conventions will come up again though, so nice to still gather opinions.

**Comment by timwoj** [comment]
I think we should move the files into a `zeek` directory at install time, but I'm not sure we need to rearrange the code to match.

**Comment by jsiwek** [comment]
> I think we should move the files into a `zeek` directory at install time, but I'm not sure we need to rearrange the code to match.

Maybe, but thinking include search paths for various usages become simpler if source/install dirs match: `#include "zeek/Foo.h"` ought to work from within Zeek source files themselves and also from external plugin code that's building against either Zeek source or install trees (unless we drop support for building against Zeek source tree?).

**Comment by rsmmr** [comment]
Agree that.a prefix `zeek/` would be preferable, seems pretty standard to organize things that way. To make things work from inside the sources files themselves, we could out symlink in place just just points to the current dir (`zeek -> .`). I'm doing that with Spicy and it seems to work pretty well.

**Comment by timwoj** [comment]
At a closer look, `Type.h` is only included by `apache/metron-bro-plugin-kafka` and it's not actually using it for anything. That entire file can be under the `detail` namespace.

**Comment by jsiwek** [comment]
Possibly some plugins are indirectly using things from `Type.h` via `Val.h` ?  A quick grep for `Type(`, I see some are creating their own `OpaqueType` and possibly inspect `FuncType` args and/or `RecordType` fields.  Also generally, the creation of a `Val` of the various non-scalar types need to be given a reference to a type.  Sometimes there's pre-created types people can use and don't need to deal with inspecting/creating via the `Type.h` API, but at least one I've seen a few times is needing to create your own `VectorType` before creating a new `VectorVal`.

Generally wonder about using "no external plugins use it" as sole criteria for `detail`ing something.  There's things we may want to actively encourage more use of even if not commonly used already?  There may also be central/core components that are heavily depended on in our own work, so may benefit the sanity/stability of our own development effort to have nicer deprecation process/hints for those?

**Comment by JonZeolla** [comment]
I've confirmed that `Type.h` is not needed by `apache/metron-bro-plugin-kafka` so I will get that removed. I'm not sure why it was there in the first place, but it seems to have been there for years.

**Comment by timwoj** [comment]
> but at least one I've seen a few times is needing to create your own VectorType before creating a new VectorVal

I saw that too and debated adding a constructor to VectorVal that just takes a TypeTag. The problem with that is that there's a couple of places where we re-use the same VectorType to make multiple VectorVals. I didn't look into the implications of not doing that.

I have a first round of namespacing on the `topic/timw/266-namespacing` branch that moves `Stmt`, `Expr`, `Anon`, `ID`, `Trigger`, and `Attr` into zeek::detail and everything in the plugin directory in to `zeek`. It also includes all of the fixes necessary to remove any deprecation warnings from the zeek code related to those moves. I'm thinking about opening a PR just to generate discussion on the approach.

**Comment by jsiwek** [comment]
> > but at least one I've seen a few times is needing to create your own VectorType before creating a new VectorVal
> 
> I saw that too and debated adding a constructor to VectorVal that just takes a TypeTag.

Other issue there is it's possibly not enough type information for vectors of non-scalars --  e.g.`TYPE_RECORD` doesn't describe an exact type, but `RecordType` does.  And at least `VectorVal::Assign()` does perform a type-check at the moment.

**Comment by timwoj** [comment]
We should be able to add anything in the base level of `analyzer`, `file_analysis`, `iosource`, `logging`, and `threading` to the `zeek` namespace. Individual plugins in those directories will need a look at on a case-by-case basis. For example the main TCP analyzer plugin should probably be in the main namespace since it's extended by others.

**Comment by timwoj** [comment]
One question is how to handle the existing namespaces that we already have. Using `file_analysis` as an example, that would become `zeek::file_analysis` for everything in the main directory that's used by other plugins. Should anything that needs to be in the detail namespace be `zeek::file_analysis::detail` or `zeek::detail::file_analysis`?

**Comment by timwoj** [comment]
Should global methods such as `type_name` from Type.h be moved into namespaces?

**Comment by jsiwek** [comment]
> One question is how to handle the existing namespaces that we already have. Using `file_analysis` as an example, that would become `zeek::file_analysis` for everything in the main directory that's used by other plugins. Should anything that needs to be in the detail namespace be `zeek::file_analysis::detail` or `zeek::detail::file_analysis`?

The one that makes more sense to me is `zeek::file_analysis::detail` since that better shows the true hierarchical relationship.  e.g. "these things are specifically an implementation detail of file analysis".

Think it also works out better in terms of allowing unqualified name lookups from within the `zeek::subspace::detail` that refer to things in the `zeek::subspace`.

> Should global methods such as `type_name` from Type.h be moved into namespaces?

I say yeah.

**Comment by timwoj** [comment]
I keep going back and forth about whether `*Type` should be in `zeek` or `zeek::detail`. I think I've landed finally on it being in `zeek`, since it's used quite frequently in pac and bif files and not just in C++ code.

**Comment by timwoj** [comment]
My previous list of classes was woefully incomplete due to my own misunderstanding of how the compiler would remove symbols during the build. Here is a more-complete list using universal-ctags and a bit of python to parse them:

https://gist.github.com/timwoj/4485ba3e63a38c7bb7844494fdfd8eaf

**Comment by rsmmr** [comment]
Yeah, I think I agree own `*Type` in `zeek.`

For sub-namespaces, I can't decide if I prefer `zeek::X::detail` or `zeek:detail::X::` so I'm fine either way.

Is it feasible to include moving the sub namespaces into `zeek::...` with the first round?

I would also suggest we include `IntrusivePtr` because that's been touch all over the place already this cycle.

**Comment by timwoj** [comment]
> For sub-namespaces, I can't decide if I prefer `zeek::X::detail` or `zeek:detail::X::` so I'm fine either way.

I think `zeek::X::detail` is the most clear.

> Is it feasible to include moving the sub namespaces into zeek::... with the first round?
> I would also suggest we include IntrusivePtr because that's been touch all over the place already this cycle.

Sure, I can work on those this week. I just pushed a commit to that PR that moves all of the Type classes to the zeek namespace.

**Comment by jsiwek** [comment]
> > Is it feasible to include moving the sub namespaces into zeek::... with the first round?
> > I would also suggest we include IntrusivePtr because that's been touch all over the place already this cycle.
> 
> Sure, I can work on those this week.

I wouldn't mind getting the chunk in #982 merged ASAP (nice work/progress btw), with sub-namespaces and `IntrusivePtr` planned for another PR sometime before 3.2.  Would you mind looking at adding the specific `IntrusivePtr` aliases, like `zeek::ValPtr`, during that move, too?

**Comment by timwoj** [comment]
How do we want to handle the move of `IntrusivePtr`? That type was added during the 3.x development cycle, so does it need a deprecated version of it to be removed in 4.1?

**Comment by timwoj** [comment]
> Would you mind looking at adding the specific IntrusivePtr aliases, like zeek::ValPtr, during that move, too?

Thoughts on naming here? For example, is simple naming like `FuncPtr` for `IntrusivePtr<Func>` enough? `FuncIntrusivePtr`?

**Comment by jsiwek** [comment]
> How do we want to handle the move of `IntrusivePtr`? That type was added during the 3.x development cycle, so does it need a deprecated version of it to be removed in 4.1?

Seems fine to just change it since my read of the release policy only describes stability/backwards-compat efforts to be in terms of LTS-to-LTS.

> > Would you mind looking at adding the specific IntrusivePtr aliases, like zeek::ValPtr, during that move, too?
> 
> Thoughts on naming here? For example, is simple naming like `FuncPtr` for `IntrusivePtr<Func>` enough?

Yeah, just appending `Ptr` to each thing is the first thing I think to do, too: `ValPtr`, `RecordValPtr`, `TypePtr`, `RecordTypePtr`, `AttrPtr`, etc.

**Comment by timwoj** [comment]
> Yeah, just appending Ptr to each thing is the first thing I think to do, too: ValPtr, RecordValPtr, TypePtr, RecordTypePtr, AttrPtr, etc.

The reason I ask is that we already have EventHandlerPtr, which is not an IntrusivePtr. But I agree I think just appending Ptr is clearest.

**Comment by timwoj** [comment]
Reopening this because there's still a lot of work to do.

**Comment by timwoj** [comment]
I ran into an actual compiler bug today: https://godbolt.org/z/RiaVzQ

gcc warns on deprecated template aliases, but clang doesn't. This is going to make moving things like List and Dict a little more difficult, because they're not going to warn on clang builds. It was reported to the clang team in 2013 but no one has picked it up to work on it: https://bugs.llvm.org/show_bug.cgi?id=17862

**Comment by timwoj** [comment]
How should the parts of the regex matcher be split up? Should just `RE_Matcher` be in the zeek namespace and all of the rest of it in detail? How about the Rule stuff?

**Comment by jsiwek** [comment]
> Should just `RE_Matcher` be in the zeek namespace and all of the rest of it in detail? How about the Rule stuff?

All that generally seems like `detail` to me (even direct `RE_Matcher` usage may be niche/uncommon, but guess we have a few places in `zeek::` API that already taking those as arguments so could see using `zeek::RE_Matcher` for that reason).

**Comment by rsmmr** [comment]
Yeah, sounds right to me.

**Comment by timwoj** [comment]
What's the opinion on everything in util? Just make it all part of `zeek` since it's mostly used by bifs and such?

**Comment by jsiwek** [comment]
A hunch is that `util.h` is a mixed bag with some things worth `detail`ing (or just re-organizing/hiding) since they're for esoteric/internal usage.  e.g. didn't look comprehensively, but at a glance, I don't know what people would care to use `out_of_memory()`, `open_package()`, `warn_if_legacy_script()` for.

**Comment by timwoj** [comment]
I talked to @rsmmr this morning. I'm going to look at splitting that file up a little bit before trying to namespace it. There's a mixed bag of semi-related functions and whatnot in there that could probably be better served by having a string_util.h, file_util.h, etc.

**Comment by timwoj** [comment]
What are your thoughts on the `bro_broker` namespace, used for everything in the `broker` directory? Changing it to `zeek::zeek_broker` seems redundant, but changing it to just `zeek::broker` means that we have to explicitly call out `::broker` for all of the non-zeek broker types and methods whenever inside the context of a zeek namespace.

**Comment by jsiwek** [comment]
Maybe `zeek::Broker` or `zeek::broker_framework` ?

**Comment by timwoj** [comment]
Does C++ differentiate namespaces based on case? That's something I don't think I've ever encountered.  `zeek::broker_framework` feels a little too close to script frameworks and might confuse someone. How about `zeek::broker_internal`? There's nothing in there that would go in a detail namespace, so that part wouldn't be redundant.

**Comment by jsiwek** [comment]
Yeah, namespace identifiers are case-sensitive, just like other identifiers.  So still suggest `zeek::Broker`: don't think the namespace identifier convention has been strictly defined as all-lowercase yet, but also personally don't mind the occasional break from convention when there's reason for it.  Or a new/alternate idea is `zeek::broker_support`.

**Comment by timwoj** [comment]
I'll go with `zeek::Broker` for now and see whether the casing feels ambiguous as I'm making changes.

**Comment by rsmmr** [comment]
One more idea: `zeek::broker_`?

**Comment by timwoj** [comment]
I just pushed a couple of commits that move the protocol analyzers. The changes I made were:

- All analyzer namespaces moved from `analyzer::*` to `zeek::analyzer:*`
- "Common" analyzers such as TCP, UDP, etc all stayed in the public namespace.
- Any class that was an actual analyzer (meaning it's called out in a Plugin.cc or used by another Plugin) stayed in the public namespace.
- Almost all enums moved to detail namespaces. There's a few exceptions, such as some of the MIME enums.
- Any Interpreter class moved to the detail namespace.

There's a few changes I need to go back and make now that I'm looking at the list of moved types (mostly moving things to `detail`), but it's mostly done.

The last things left are:
- files like util.h and Net.h that are just a collection of functions, not types per se. Net.h is easy, since it can just all be in zeek::detail. util.h is harder.
- Everything in zeekygen. I'm debating whether to move that to zeek::zeekygen, which feels really redundant.

**Comment by jsiwek** [comment]
> * Everything in zeekygen. I'm debating whether to move that to zeek::zeekygen, which feels really redundant.

Does initially seem redundant, but with `zeek::` versus `zeek::detail::` being used to signal "API stability", then still makes sense to namespace `zeekygen` stuff for clarity.  Also I'd go with `zeek::detail::zeekygen` since it's not originally intended for public consumption and also don't expect demand to use it directly anyway.

**Comment by timwoj** [comment]
Alright, I think this is the last question for this task. The `Plugin.cc` file for every plugin contains a namespace. Should those all be changed to `zeek::plugin::<something>`? That shouldn't cause any issues with plugin loading or anything like that, correct?

**Comment by rsmmr** [comment]
> Should those all be changed to `zeek::plugin::<something>`? 

yes

> That shouldn't cause any issues with plugin loading or anything like that, correct?

No, can't think of anything that would break.

**Comment by vpax** [comment]
Having poked my nose into some of @timwoj 's recent commits, I have to say that some of the namespace stuff is really clunkifying the code to my eye.  Chatting w/ @jsiwek I'm wondering about, at a minimum, switching to `using namespace zeek` inside `.cc` files so they're easier to scan.

**Comment by timwoj** [comment]
> Chatting w/ @jsiwek I'm wondering about, at a minimum, switching to using namespace zeek inside .cc files so they're easier to scan.

I believe I tried that at the start of this work and ran into issues where the original name is still defined in the global namespace (for backwards compat until 4.1). The compiler complains about ambiguity between the namespaced symbol and the non-namespaced one. I can certainly revisit it though.

**Comment by timwoj** [comment]
I went back and took the namespace wrapper out of `Val.cc` and replaced it with `using namespace zeek`:

```
[18/23] Building CXX object src/CMakeFiles/zeek_objs.dir/Val.cc.o
FAILED: src/CMakeFiles/zeek_objs.dir/Val.cc.o 
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -DDOCTEST_CONFIG_SUPER_FAST_ASSERTS -Isrc -I../src -I../auxil/highwayhash -I/usr/local/opt/openssl/include -Iauxil/binpac/lib -I../auxil/binpac/lib -I../auxil/rapidjson/include -I../auxil/paraglob/include -Iauxil/broker/caf/libcaf_core -I../auxil/broker/caf/libcaf_test -I../auxil/broker/caf/libcaf_openssl -I../auxil/broker/caf/libcaf_io -I../auxil/broker/caf/libcaf_core -I../auxil/broker/caf/libcaf_opencl -Iauxil/broker/include -I../auxil/broker/include -I. -Wall -Wno-unused -Wno-register -Werror=vla -std=c++17 -O3 -DNDEBUG -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -MD -MT src/CMakeFiles/zeek_objs.dir/Val.cc.o -MF src/CMakeFiles/zeek_objs.dir/Val.cc.o.d -o src/CMakeFiles/zeek_objs.dir/Val.cc.o -c ../src/Val.cc
../src/Val.cc:51:4: warning: 'Val' is deprecated: Remove in v4.1. Use zeek::Val instead. [-Wdeprecated-declarations]
Val::Val(zeek::FuncPtr f)
   ^
../src/Trigger.h:14:1: note: 'Val' has been explicitly marked deprecated here
ZEEK_FORWARD_DECLARE_NAMESPACED(Val, zeek);
^
./zeek-config.h:282:89: note: expanded from macro 'ZEEK_FORWARD_DECLARE_NAMESPACED'
#define ZEEK_FORWARD_DECLARE_NAMESPACED( ... ) GET_MACRO(_0, ##__VA_ARGS__, ZEEK_FDN_3, ZEEK_FDN_2, ZEEK_FDN_1, ZEEK_FDN_0)(__VA_ARGS__)
                                                                                        ^
../src/Val.cc:58:30: error: reference to 'base_type' is ambiguous
                = make_intrusive<FileType>(base_type(TYPE_STRING));
                                           ^
../src/Type.h:928:16: note: candidate found by name lookup is 'base_type'
constexpr auto base_type [[deprecated("Remove in v4.1. Use zeek::base_type instead.")]] = zeek::base_type;
               ^
../src/Type.h:872:16: note: candidate found by name lookup is 'zeek::base_type'
const TypePtr& base_type(zeek::TypeTag tag);
               ^
```

That's just the first bit of it. Lots more ambiguity follows from that. I agree that the extra namespace declarations and having to be very specific with where a symbol comes from add quite a lot of clutter (especially in function definitions). 

One thing we could do is go through all of the places where a namespace wraps a block of code and remove direct uses of the namespace inside of that block. That works because the namespace wrapper is explicit about what namespace you're in, and it will try to find the symbol from that namespace first. I couldn't do that as I was working on it because I was chipping away at parts of the code base instead of doing it all at once. For example in both Val.h and Val.cc, I can remove all of the specific uses of `zeek::` since those files are wrapped in `namespace zeek` blocks.

That said, once the deprecated symbols are removed in v4.1 the ambiguity goes away and we can clean up a bit where the ambiguity was breaking things. That has to wait until after 4.0 thought unfortunately.

**Comment by jsiwek** [comment]
> One thing we could do is go through all of the places where a namespace wraps a block of code and remove direct uses of the namespace inside of that block.

That sounds fine to me, plus could think about adding `namespace zeek { }` blocks in certain well-traveled places that don't already have it.  The underlying problem that is worth addressing is that the use of fully-qualified `zeek::` and `zeek::detail::` names adds friction to reading/writing code.  Whether we're able to use unqualified names via `using namespace foo;` or defining things inside `namespace foo { }` blocks probably doesn't matter so much besides style/convention preferences or just the current reality that deprecated symbols prevent doing the form unambiguously.  (Header files would not have the option of going the `using` path anyway).

For timelines, I'd find it helpful if there's any improvement that can be made before 4.0.

**Comment by timwoj** [comment]
Alright, I'll tackle that after the part 4 PR lands so that I can get all of it done at once.

**Comment by jsiwek** [comment]
Anything still planned for this issue or can it be closed?

**Comment by timwoj** [comment]
Nope, this one is (finally) done.

**Comment by JonZeolla** [comment]
For plugin maintainers, where can we look for the updated documentation or migration path?

**Comment by rsmmr** [comment]
The NEWS file is summarizing changes pretty comprehensively. If you find anything missing, or are looking for something different, let us know.
